// orbit-core/dist/cjs/transform/operators.js
'use strict';

exports.addRecord = addRecord;
exports.replaceRecord = replaceRecord;
exports.removeRecord = removeRecord;
exports.replaceKey = replaceKey;
exports.replaceAttribute = replaceAttribute;
exports.addToHasMany = addToHasMany;
exports.removeFromHasMany = removeFromHasMany;
exports.replaceHasMany = replaceHasMany;
exports.replaceHasOne = replaceHasOne;

function addRecord(record) {
  return { op: 'addRecord', record: record };
}

function replaceRecord(record) {
  return { op: 'replaceRecord', record: record };
}

function removeRecord(record) {
  return { op: 'removeRecord', record: record };
}

function replaceKey(record, key, value) {
  return { op: 'replaceKey', record: record, key: key, value: value };
}

function replaceAttribute(record, attribute, value) {
  return { op: 'replaceAttribute', record: record, attribute: attribute, value: value };
}

function addToHasMany(record, relationship, relatedRecord) {
  return { op: 'addToHasMany', record: record, relationship: relationship, relatedRecord: relatedRecord };
}

function removeFromHasMany(record, relationship, relatedRecord) {
  return { op: 'removeFromHasMany', record: record, relationship: relationship, relatedRecord: relatedRecord };
}

function replaceHasMany(record, relationship, relatedRecords) {
  return { op: 'replaceHasMany', record: record, relationship: relationship, relatedRecords: relatedRecords };
}

function replaceHasOne(record, relationship, relatedRecord) {
  return { op: 'replaceHasOne', record: record, relationship: relationship, relatedRecord: relatedRecord };
}
